libs-gui/Documentation/gopen.1
.\"gopen(1) man page
.\"put together by Martin Brecher <martin@gnustep.de>
.\"Copyright (C) 2005 Free Software Foundation, Inc.
.\"
.\"Process this file with
.\"groff -man -Tascii gopen.1
.\"
.TH GOPEN 1 "August 2003" GNUstep "GNUstep System Manual"
.SH NAME
gopen \- open files
.SH SYNOPSIS
.B gopen
.RB [ -a 
.IR application ]
.RB [ -o ]
.RB [ -p ]
.RB [ -NXHost 
.IR hostname ]
.RI [ filename ]
.P
.SH DESCRIPTION
The 
.B gopen
command allows you open a file (or directory) as if you had double clicked
the object's icon.
.PP
It is possible to specify one or more 
.I filenames
which are interpreted relative to the current working directory. 
.PP
By default, 
.B gopen
will open
.I filename
with the application currently assigned to the file's extension. But by
specifying the 
.B -a
flag on the command line you can tell
.B gopen
to open the file with another application.
.PP
You can also directly print a file without (
.B -p
) or with (
.B -p -o 
) opening it.
.PP
As with most other GNUstep software, it is easily possible
to attach the process to a remote window server using the
.B -NXHost
flag.
.PP
.SH OPTIONS
.IP "\fB-a \fIapplication"
use
.I application
to open
.I filename
.IP "\fB-o"
open 
.I filename
(may be used in conjunction with
.B -p
).
.IP "\fB-p"
print
.I filename
instead of opening.
.IP "\fB-NXHost \fIhostname"
attach to remote window server on
.I hostname
.PP
.SH EXAMPLES
Open all files with a .txt extension in the current directory:
.PP
.I gopen *.txt
.PP
To open a source file in CodeEditor (instead of the current default 
application), type:
.PP
.I gopen -a CodeEditor MySourceFile.m
.PP
To print the .plan file in your home folder, use the
.B -p
flag:
.PP
.I gopen -p ~/.plan
.PP
.SH BUGS
Using the
.B -a
.I application
argument,
.B gopen
allows you to open any kind of file with any application you want. This does,
however, not work with applications employing the NSDocument architecture
as they will only receive filenames whose types the application has registered.
To circumvent this limitation, applications may advertise their ability to open
all kinds of files through the "*" filetype.
When
.B gopen
is used with only the 
.B -a
.I application 
argument but no
.I filename
 present, the application is simply launched without opening a specific file.
.P
.SH SEE ALSO
GNUstep(7), openapp(1)
.P
.SH HISTORY
.B gopen
was written November 2001.
.P
This manual page was first written July 2003.
.P
.SH AUTHORS
.B gopen 
was written by Gregory Casamento <greg_casamento@yahoo.com>.
.P
This man page was written by Martin Brecher <martin@mb-itconsulting.com>.


libs-gui/NEWS
1 NEWS
******

The currently released version of the library is '0.28.0'.

1.1 Noteworthy changes in version '0.28.0'
==========================================

This version adds support for modern XIB files and many new classes.
Plus the usual bunch of bug fixes.

   * Support loading of document XIB files.
   * Improve Key Value Binding for NSArrayController and add more
     bindings.
   * Better support for multi monitor usage and other improvement in the
     backend integration.
   * Add classes NSFontCollection, NSColorSampler, NSSpeechRecognizer,
     NSAppearance, NSPDFInfo, NSPICTImageRep, NSCIImageRep,
     NSPDFImageRep, NSPDFPanel, NSDataAsset, NSDatePicker,
     NSDatePickerCell, NSPredicateEditor, NSPredicateEditorRowTemplate,
     NSRuleEditor, NSGestureRecognizer, NSButtonTouchBarItem,
     NSCandidateListTouchBarItem, NSClickGestureRecognizer,
     NSColorPickerTouchBarItem, NSCustomTouchBarItem,
     NSGroupTouchBarItem, NSMagnificationGestureRecognizer,
     NSPanGestureRecognizer, NSPickerTouchBarItem,
     NSPopoverTouchBarItem, NSPressGestureRecognizer,
     NSRotationGestureRecognizer, NSSharingServicePickerTouchBarItem,
     NSSliderTouchBarItem, NSStepperTouchBarItem, NSTouchBarItem,
     NSTouchBar, NSTouch, NSDockTile.
   * Implement NSEPSImageRep.
   * Better encoding handling in RTF files.
   * Theming and drawing improvements.
   * Increase small font size to 10.
   * New cursor and stepper images.
   * Move NSFileWrapper to Foundation.
   * Fixed build on Debian GNU/kFreeBSD.
   * With command line argument -autolaunch YES, do not activate the
     application when -activateIgnoringOtherApps: is invoked.
   * Improvements to WindowMaker compatibility (e.g.  WMFHideApplication
     support).
   * Lowered NSFloatingWindowLevel by one to distinguish floating panels
     from menus.

1.2 Noteworthy changes in version '0.27.0'
==========================================

This version includes numerous bugfixes, compatibility improvements and
other changes accumulated over the last year.  It also enables work to
be done on integrating NSViews with a Core Animation renderer.

   * Make targetForAction safer.
   * Speed up menu updates.
   * Clean up speech tool compilation and switch to newer interface.
   * Fix bug in CUPS subclassing introduced in last release.
   * Minor improvements to typesetting.
   * Add NSIsControllerMarker.
   * Fix tracking on segmented cell.
   * Bring slider cell closer to Cocoa implementation.
   * Add ivar for Core Animation in NSView.
   * Improve border calculation on printing.
   * Lazy load app icon.
   * Better detection of removable volumes.
   * Polish translations.
   * Japanese translations.
   * Lots of bug fixes.

1.3 Noteworthy changes in version '0.26.2'
==========================================

This version is a small, but important bugfix release.

   * printing: Fix allocation of the CUPS printing classes.
   * installation: Fix the configure script.

1.4 Noteworthy changes in version '0.26.1'
==========================================

This version is released to conincide with version 1.25.1 of
gnustep-base, which contains changes required for this version of
gnustep-gui and gnustep-back.

   It includes an important workaround for users of GNUstep Objective-C
Runtime (libobjc2) and non-fragile ABI to avoid a bug in interaction
between the clang compiler and the runtime when non-fragile ABI is in
use.  Specifically, Clang and the runtime may disagree on what is the
offset of an ivar in a class's RAM. This manifested in a crash at
application startup due to misalignment of _gcontext inside NSThread.
See the mailing list discussion
(http://lists.gnu.org/archive/html/discuss-gnustep/2017-12/msg00129.html)
for more information.

   It also contains the following changes:

   * tests: Cleanup of warnings.
   * tests: Fix text system deallocation test.
   * printing: Undefine __BLOCKS__ before including cups.h, as some
     versions of the header expect that libdispatch is present and used
     if __BLOCKS__ is defined.
   * graphics context: Workaround for Clang+libobjc2+nonfragile ABI
     issue.

1.5 Noteworthy changes in version '0.26.0'
==========================================

This version was bumped due to previous binary incompatibilities between
0.25.0 and 0.25.1.  This version also includes numerous compatibility
improvements from the Summer of Code project, and a wide variety of
other fixes.  Notably, it fixes the use of cupsGetPPD() in the printing
system.

   * printing: Add an include to get deprecated function cupsGetPPD() on
     newer CUPS systems.
   * chore: Bump required base version.
   * tiff: Support for writing resolution.
   * jpeg: Save resolution information if it is different from 72 dpi.
   * save panel: Fix return type of sorting function.
   * events: Add some newer Cocoa enums and one method with dummy
     implementation.
   * speech synthesis: NSSpeechSynthesizerDelegate is now a @protocol on
     runtimes that support it.
   * pasteboard: New type identifiers.
   * translations: Some work on Polish, Russian and German translations
   * cell: Improvements to mouse tracking logic on NSCell.
   * image: If an unknown named image is unarchived with a coder or
     keyed coder, keep the name.
   * screen: Add -backingScaleFactor and return 1.0.
   * window: Return 1.0 from -backingScaleFactor.

   * compatibility: Numerous stub implementations of constants, classes
     and methods to improve source-level compatibility.
   * other bugfixes

1.6 Noteworthy changes in version '0.25.1'
==========================================

   * JPEG (saving) alpha channel fixes and size with resolution != 72
   * JPEG resolution read support
   * TIFF saving fixes
   * Improved volumes mounting and support
   * Portability improvements in volume mounting and support
   * Corrected layout of empty strings
   * Only update visible menus

1.7 Noteworthy changes in version '0.25.0'

libs-gui/Tools/gopen.m
/* This tool opens the appropriate application from the command line
   based on what type of file is being accessed. 

   Copyright (C) 2001 Free Software Foundation, Inc.

   Written by:  Gregory Casamento <greg_casamento@yahoo.com>
   Created: November 2001

   This file is part of the GNUstep Project

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License
   as published by the Free Software Foundation; either version 3
   of the License, or (at your option) any later version.
    
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public  
   License along with this library; see the file COPYING.
   If not, see <http://www.gnu.org/licenses/> or write to the 
   Free Software Foundation, 51 Franklin Street, Fifth Floor, 
   Boston, MA 02110-1301, USA.
*/

#include <Foundation/NSArray.h>
#include <Foundation/NSData.h>
#include <Foundation/NSDictionary.h>
#include <Foundation/NSString.h>
#include <Foundation/NSProcessInfo.h>
#include <Foundation/NSTask.h>
#include <Foundation/NSAutoreleasePool.h>
#include <Foundation/NSException.h>
#include <Foundation/NSFileManager.h>
#include <Foundation/NSFileHandle.h>
#include <Foundation/NSPathUtilities.h>
#include <Foundation/NSURL.h>
#include <Foundation/NSValue.h>
#include <Foundation/NSUserDefaults.h>
#include <AppKit/NSApplication.h>
#include <AppKit/NSWorkspace.h>

static NSString*
absolutePath(NSFileManager *fm, NSString *path)
{
  path = [path stringByStandardizingPath];
  if ([path isAbsolutePath] == NO)
    {
      path = [[fm currentDirectoryPath] stringByAppendingPathComponent: path];
    }
  return path;
}

int
main(int argc, char** argv, char **env_c)
{
  CREATE_AUTORELEASE_POOL(pool);
  NSEnumerator  *argEnumerator = nil;
  NSWorkspace   *workspace = nil;
  NSFileManager *fm = nil;
  NSString      *arg = nil;
  NSString      *editor = nil;
  NSString      *terminal = nil;
  NSString      *application = nil;
  NSString      *filetoopen = nil;
  NSString      *filetoprint = nil;
  NSString      *nxhost = nil;
  BOOL		isDir;
  BOOL		exists;
  NSURL		*u;

#ifdef GS_PASS_ARGUMENTS
  [NSProcessInfo initializeWithArguments:argv count:argc environment:env_c];
#endif
  argEnumerator = [[[NSProcessInfo processInfo] arguments] objectEnumerator];
  workspace = [NSWorkspace sharedWorkspace];
  fm = [NSFileManager defaultManager];
  
  // Default applications for opening unregistered file types....
  editor = [[NSUserDefaults standardUserDefaults]
    stringForKey: @"GSDefaultEditor"];
  terminal = [[NSUserDefaults standardUserDefaults]
    stringForKey: @"GSDefaultTerminal"];

  // Process options...
  application = [[NSUserDefaults standardUserDefaults] stringForKey: @"a"];
  filetoopen = [[NSUserDefaults standardUserDefaults] stringForKey: @"o"];
  filetoprint = [[NSUserDefaults standardUserDefaults] stringForKey: @"p"];
  nxhost = [[NSUserDefaults standardUserDefaults] stringForKey: @"NXHost"];

  if (application)
    {
      /* Start the application now,
       * later on we will use it for file opening. 
       */
      [workspace launchApplication: application];
    }
  
  if (filetoopen)
    {
      exists = [fm fileExistsAtPath: arg isDirectory: &isDir];
      if (exists == NO)
	{
	  if ([filetoopen hasPrefix: @"/"] == NO
	    && (u = [NSURL URLWithString: filetoopen]) != nil)
	    {
	      [workspace openURL: u];
	    }
	}
      else
	{
	  filetoopen = absolutePath(fm, filetoopen);
	  [workspace openFile: filetoopen
	      withApplication: application];
	}
    }

  if (filetoprint)
    {
      filetoprint = absolutePath(fm, filetoprint);
      puts("Not implemented");
    }

  if (nxhost)
    {
      puts("Not implemented");
    }

  if (argc == 1)
    {
      NSFileHandle	*fh = [NSFileHandle fileHandleWithStandardInput];
      NSData		*data = [fh readDataToEndOfFile];
      NSString		*tempFile;
      int		processId;

      tempFile = NSTemporaryDirectory();
      tempFile = [tempFile stringByAppendingPathComponent: @"openfiletmp"];
      processId = [[NSProcessInfo processInfo] processIdentifier];
      tempFile = [tempFile stringByAppendingFormat: @"%d", processId];
      tempFile = [tempFile stringByAppendingString: @".txt"];
      [data writeToFile: tempFile atomically: YES];
      [workspace openFile: tempFile withApplication: editor];      
    }

  [argEnumerator nextObject]; // skip the first element, which is empty.
  while ((arg = [argEnumerator nextObject]) != nil)
    {
      NSString *ext = [arg pathExtension];
      
      if ([arg isEqualToString: @"-a"]
        || [arg isEqualToString: @"-o"]
        || [arg isEqualToString: @"-NXHost"])
	{
	  // skip since this is handled above...
	  arg = [argEnumerator nextObject];
	}
      else // no option specified
	{
	  NS_DURING
	    {
	      exists = [fm fileExistsAtPath: arg isDirectory: &isDir];
	      if (exists == YES)
		{
		  arg = absolutePath(fm, arg);
		  if (isDir == NO && [fm isExecutableFileAtPath: arg])
		    {
		      [workspace openFile: arg withApplication: terminal];
		    }
		  else // no argument specified
		    {
		      // First check to see if it's an application
		      if ([ext isEqualToString: @"app"]
			|| [ext isEqualToString: @"debug"]
			|| [ext isEqualToString: @"profile"])
			{
			  [workspace launchApplication: arg];
			}
		      else
			{
			  if (![workspace openFile: arg
				   withApplication: application])
			    {
			      // no recognized extension,
			      // run application indicated by environment var.
			      NSLog(@"Opening %@ with %@",arg,editor);
			      [workspace openFile: arg withApplication: editor];
			    }
			}
		    }
		}
	      else if ([arg hasPrefix: @"/"] == NO
		&& (u = [NSURL URLWithString: arg]) != nil
		&& [u scheme] != nil)
		{
		  [workspace openURL: u];
		}
	      else 
		{
		  arg = absolutePath(fm, arg);
		  GSPrintf(stdout, @"The file %@ does not exist.\n", arg);
		}
	    }
	  NS_HANDLER
	    {
	      NSLog(@"Exception while attempting open file %@ - %@: %@",
		    arg, [localException name], [localException reason]);
	    }
	  NS_ENDHANDLER
	}
    }
  RELEASE(pool);
  exit(EXIT_SUCCESS);
}
